runtime/src/iree/base/internal/math.h

```diff
@@ -400,16 +421,23 @@ static inline uint32_t iree_math_truncate_f32_to_bits_rounding_to_nearest_even(
     } else if (arithmetic_exp + dst_exp_bias <= 0) {
       // Underflow. Generate a subnormal or zero.
       dst_exp = 0;
+      // Arithmetic exponent of destination type subnormals.
+      int dst_arithmetic_exp = 1 - dst_exp_bias;
       // The exponent has to be clamped to 0 when the value
       // (arithmetic_exp + dst_exp_bias) is negative. This has to be compensated
       // by right-shifting the subnormal mantissa.
-      int exp_to_encode_as_bitshift = -(arithmetic_exp + dst_exp_bias);
-      int shift_amount =
-          f32_mantissa_bits - dst_mantissa_bits + exp_to_encode_as_bitshift;
-      if (shift_amount >= f32_mantissa_bits) {
+      int shift_amount = f32_mantissa_bits - dst_mantissa_bits -
+                         arithmetic_exp + dst_arithmetic_exp;
+      if (shift_amount < 0 || shift_amount >= f32_mantissa_bits) {
         dst_mantissa = 0;
       } else {
-        dst_mantissa = f32_mantissa >> shift_amount;
+        // Source f32 value is normal so has an implied 1... leading bit.
```

Comment:
Unless this code is already not applicable to f32-to-bf16, I think the logic here is wrong and might flush f32 denormals to 0